File: solutions/322/main.py
from typing import List 

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # dp table
        MAX = amount + 1
        dp = [[MAX] * (amount+1) for _ in range(len(coins)+1)]
        # initialize
        dp[0][0] = 0
        for i in range(1, len(coins)+1):
            # note, this is NOT for j in range(1, amount+1):
            for j in range(amount+1):
                if j < coins[i-1]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-coins[i-1]] + 1)
        return dp[-1][-1] if dp[-1][-1] < MAX else -1

if __name__ == "__main__":
    coins = [1,2,5] 
    amount = 11
    s = Solution()
    assert s.coinChange(coins, amount) == 3

File: solutions/322/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Tue Nov 23 15:20:35 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // dp table
        const int MAX = amount + 1;
        vector<int> dp(amount + 1, MAX);
        // initialize
        dp[0] = 0;
        for (int i = 0; i < amount + 1; ++i) {
            for (int j = 0; j < coins.size(); ++j) {
                // dont forget +1
                if (i >= coins[j]) dp[i] = min(dp[i], dp[i-coins[j]]+1);
            }
        }
        return dp[amount] < MAX ? dp[amount] : -1;
    }
};

TEST(Test322, SimpleTest) {
    vector<int> coins{
        1,2,5
    };
    const int amount = 11; 
    Solution s; 
    EXPECT_EQ(s.coinChange(coins, amount), 3); 
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS();
}
